### ISSUE TITLE
Strange Metrics/AbcSize behaviour

### ISSUE THREAD

**Issue by dvandersluis** [rubocop-hq/rubocop#7563]
Consider the following class:

```ruby
class Tokenizer
  def tokenize
    scanner = StringScanner.new(string)

    until scanner.eos?
      scanner.skip(/\s+/)

      tokens << if scanner.scan(regexp1)
        Token1.new(scanner.matched)
      elsif scanner.scan(regexp2)
        Token2.new(scanner.matched)
      elsif scanner.scan(regexp3)
        Token3.new(scanner.matched)
      elsif scanner.scan(regexp4)
        Token4.new(scanner.matched)
      else
        scanner.scan(regexp5)
        Token5.new(scanner.matched)
      end
    end
  end
end
```

The `tokenize` method raises a `Metrics/AbcSize` offense:

```
Metrics/AbcSize: Assignment Branch Condition size for tokenize is too high. [<1, 26, 6> 26.7/15]
```

However, if I extract `scanner` to its own method, as below:

```ruby
class Tokenizer
  def scanner
    @scanner ||= StringScanner.new(string)
  end

  def tokenize
    until scanner.eos?
      # ... same as above, clipped for space
    end
  end
end
```

the ABC score actually *increases!*

```
Metrics/AbcSize: Assignment Branch Condition size for tokenize is too high. [<0, 36, 6> 36.5/15]
```

## Expected behavior

`AbcSize` should be reduced by removing code

## Actual behavior

Extracting code to another method increased the number of branches calculated by 10.

## RuboCop version

```
0.77.0 (using Parser 2.6.5.0, running on ruby 2.6.5 x86_64-darwin18)
```

**Comment by buehmann** [comment]
The score in the B (branch) component increases because after your extraction every occurrence of `scanner` in the `tokenize` method has become a method call. There are 12 such new calls, and you remove two others (`new` and `string`). That's why the B score goes up from 26 to 36.

**Comment by dvandersluis** [comment]
Thanks @buehmann! I still think that the result was unexpected but your explanation makes sense. 👍

**Comment by ixti** [comment]
I think AbcSize cop counts branches over-aggressively:

``` ruby
def example
  result = SomeCommand.call(:a => 1)

  if result.failure?
    warn { "... #{result.errors.first}" }
    return head :bad_request
  end

  send_data SomeMessage.new({
    :id    => result[:record].id,
    :name  => result[:record].name,
    :token => result[:token].to_s
  }).encode
end
```

AbcSize cop sees here 15 branches. IMO treating `result[]` as branching is arguable here. I might be wrong though. :D After thinking about this code again I just realized that probably `result` should provide `SomeMessage`, so, I don't know :D

**Comment by Drenmi** [comment]
_Technically_ `#[]` is a method call, but I think a case might be made that it should perhaps be considered an exception, considering that it's primarily used for hash- and array indexing in Ruby. I haven't given it enough thought to have an opinion on it.

**Comment by ixti** [comment]
@Drenmi Yeah, I understand that `#[]` is a method call. It's just I was a bit confused as it was just a Hash at first. But don't have strong opinion either. :D

**Comment by CandleCandle** [comment]
Another example of what I would consider over aggressive 'branch' counting:

```ruby
def check_lengths(bytes_as_string)
  offset  = 0
  bytes = bytes_as_string.unpack('C*')
  while offset < bytes.length - 1 do
    offset += bytes[offset + 0] << 24 |
              bytes[offset + 1] << 16 |
              bytes[offset + 2] << 8 |
              bytes[offset + 3]
  end
  raise "verification failure" if offset  != bytes.length
end
```
Rough counting by hand: 1x unpack; 4x array index; 4x addition; 4x bit shift; 3x bitwise or. That's 16 "branches" according to `evaluate_branch_nodes` [0].

Is it worth adding another set of ignorable functions (e.g. arithmetic operators, array access)

[0] https://github.com/rubocop-hq/rubocop/blob/6940bb578273bf993e4af7cf8abc19d64490a676/lib/rubocop/cop/metrics/utils/abc_size_calculator.rb#L52

**Comment by marcandre** [comment]
For anyone stumbling on this issue: the config `CountRepeatedAttributes` resolves this issue *and* would count the repeated `scanner.matched` as a single branch too.
